compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs

```diff
@@ -422,13 +424,11 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
 
                 ty::Adt(def, args) if def.is_struct() => match def.non_enum_variant().tail_opt() {
                     None => tcx.types.unit,
-                    Some(field_def) => {
-                        let self_ty = field_def.ty(tcx, args);
+                    Some(tail_def) => {
+                        let tail_ty = tail_def.ty(tcx, args);
+                        let predicate = goal.predicate.with_self_ty(tcx, tail_ty).as_projection();
```

Comment:
instead of adding it as a nested goal, simply return the `<tail_ty as Pointee>::Metadata`. If the caller needs a fully normalized type, they call `NormalizesTo` in a loop.

I think you're using `as_projection()` here as otherwise a tail whose metadata should remain rigid would result in a n error. However, `NormalizesTo` in the new solver has to only normalize by 1 step. This differs from the old solver where `project` has to deeply normalize.